src/main/java/net/neoforged/neoforge/common/extensions/IBlockStateExtension.java

```diff
@@ -129,11 +129,35 @@ default boolean canHarvestBlock(BlockGetter level, BlockPos pos, Player player)
      * @param willHarvest The result of {@link #canHarvestBlock}, if called on the server by a non-creative player, otherwise always false.
      * @param fluid       The current fluid and block state for the position in the level.
      * @return True if the block is actually destroyed.
+     * @deprecated Use the tool stack sensitive version bellow.
      */
+    @Deprecated(forRemoval = true, since = "1.21.8")
     default boolean onDestroyedByPlayer(Level level, BlockPos pos, Player player, boolean willHarvest, FluidState fluid) {
         return self().getBlock().onDestroyedByPlayer(self(), level, pos, player, willHarvest, fluid);
     }
 
+    /**
+     * Called when a player removes a block. This is responsible for
+     * actually destroying the block, and the block is intact at time of call.
+     * This is called regardless of whether the player can harvest the block or
+     * not.
+     *
+     * Return true if the block is actually destroyed.
+     *
+     * This function is called on both the logical client and logical server.
+     *
+     * @param level       The current level
+     * @param player      The player damaging the block, may be null
+     * @param toolStack   The tool the player used to destroy the block.
```

Comment:
I would probably be a bit more descriptive here or in the description and say the `toolStack` is the stack prior to mining the block. If they want to get the state of the current mainhand, then use whatever the player method is.

Reply:
That's good enough for the parameter description. I don't think we want users modifying the actual stack here, instead overriding `Item#mineBlock`. If we are, then there should also be another line in the description for that.